core-bundle/src/Resources/contao/classes/FrontendTemplate.php

```diff
@@ -348,32 +348,55 @@ private function setCacheHeaders(Response $response)
 		/** @var $objPage PageModel */
 		global $objPage;
 
+		// Do not cache response if caching was not configured at all and make sure the response is private.
 		if (($objPage->cache === false || $objPage->cache < 1) && ($objPage->clientCache === false || $objPage->clientCache < 1))
 		{
-			$response->headers->addCacheControlDirective('no-cache');
-			$response->headers->addCacheControlDirective('no-store');
+			$response->headers->set('Cache-Control', 'no-cache, no-store');
 
 			return $response->setPrivate();
 		}
 
-		// Do not cache the response if a user is logged in or the page is protected
-		// TODO: Add support for proxies so they can vary on member context
-		if (FE_USER_LOGGED_IN === true || BE_USER_LOGGED_IN === true || $objPage->protected || $this->hasAuthenticatedBackendUser())
+		// Do not cache response if the response contains cookies or the current request is considered private
+		if (0 !== \count($response->headers->getCookies()) || $this->isCurrentRequestPrivate())
 		{
-			$response->headers->addCacheControlDirective('no-cache');
-			$response->headers->addCacheControlDirective('no-store');
+			$response->headers->set('Cache-Control', 'no-cache, no-store');
 
 			return $response->setPrivate();
```

Comment:
I think the `$objPage->clientCache` should also be applied here. If I configure a page in the tree to be user-cacheable, it can still be customized, right?